src/sources/timer.rs

```diff
@@ -0,0 +1,538 @@
+    counter: u32,
+    deadline: Instant,
+}
+
+impl Timer {
+    /// Create a timer that will fire immediately when inserted in the event loop
+    pub fn immediate() -> Timer {
+        Self::from_deadline(Instant::now())
+    }
+
+    /// Create a timer that will fire after a given duration from now
+    pub fn from_duration(duration: Duration) -> Timer {
+        Self::from_deadline(Instant::now() + duration)
+    }
+
+    /// Create a timer that will fire at a given instant
+    pub fn from_deadline(deadline: Instant) -> Timer {
+        Timer {
+            registration: None,
+            counter: 0xFFFFFFFF,
+            deadline,
+        }
+    }
+}
+
+impl EventSource for Timer {
+    type Event = Instant;
+    type Metadata = ();
+    type Ret = TimeoutAction;
+    type Error = std::io::Error;
+
+    fn process_events<F>(
+        &mut self,
+        _: Readiness,
+        token: Token,
+        mut callback: F,
+    ) -> Result<PostAction, Self::Error>
+    where
+        F: FnMut(Self::Event, &mut Self::Metadata) -> Self::Ret,
+    {
+        if let Some((my_token, ref wheel)) = self.registration {
+            if my_token != token {
+                return Ok(PostAction::Continue);
+            }
+            let new_deadline = match callback(self.deadline, &mut ()) {
+                TimeoutAction::Drop => return Ok(PostAction::Remove),
+                TimeoutAction::ToInstant(instant) => instant,
+                TimeoutAction::ToDuration(duration) => Instant::now() + duration,
+            };
+            // If we received an event, we MUST have a valid counter value
+            wheel
+                .lock()
+                .unwrap()
+                .insert_reuse(self.counter, new_deadline, my_token);
+            self.deadline = new_deadline;
+        }
+        Ok(PostAction::Continue)
+    }
+
+    fn register(&mut self, poll: &mut Poll, token_factory: &mut TokenFactory) -> crate::Result<()> {
+        let wheel = poll.timers.clone();
+        let token = token_factory.token();
+        self.counter = wheel.lock().unwrap().insert(self.deadline, token);
+        self.registration = Some((token, wheel));
+        Ok(())
+    }
+
+    fn reregister(
+        &mut self,
+        poll: &mut Poll,
+        token_factory: &mut TokenFactory,
+    ) -> crate::Result<()> {
+        self.unregister(poll)?;
+        self.register(poll, token_factory)
+    }
+
+    fn unregister(&mut self, poll: &mut Poll) -> crate::Result<()> {
+        poll.timers.lock().unwrap().cancel(self.counter);
+        self.registration = None;
+        self.counter = 0xFFFFFFFF;
+        Ok(())
+    }
+}
+
+/// Action to reschedule a timeout if necessary
+#[derive(Debug)]
+pub enum TimeoutAction {
+    /// Don't reschedule this timer
+    Drop,
+    /// Reschedule this timer to a given [`Instant`]
+    ToInstant(Instant),
+    /// Reschedule this timer to a given [`Duration`] in the future
+    ToDuration(Duration),
+}
+
```

Comment:
I'm still struggling to relate the different parts of this (`TimeoutData`, `TimerWheel`, the `timerfd` in the `Poll`) and how they interact with each other. Would you be able to write a short (perhaps non-doc) comment block that explains it? Even just some bullet points would help (me, and probably future contributors too).